### ISSUE TITLE
Synchronize Haskell version to new Changes

### ISSUE THREAD

**Issue by starcraftman** [starcraftman/zsh-git-prompt#5]
I will maintain a list here of things I need added to the Haskell version to achieve parity with Python. So far:
- Stashed support, should come out last number from gitstatus. Fetchable with: `git stash list` . Just count number of lines.

**Comment by starcraftman** [comment]
TODO - Make an explicit list of changes of expected behaviour of Haskell gitstatus output.

**Comment by starcraftman** [comment]
![image](https://user-images.githubusercontent.com/470400/40859986-a7b1d304-65b1-11e8-8a56-42490175d5ba.png)

Saved for later, to demonstrate new feature.

**Comment by starcraftman** [comment]
I've updated the interface per-emptively, will now merge support for rebasing/merging indicators onto the project. As with other features, these remain python only until someone makes a PR for support or I use some time to learn Haskell better.

**Comment by madnight** [comment]
@starcraftman yup, i'll implement all of this features. Yesterday i implemented the feature that it is now possible to run the Haskell version without stdin pipe and i found a small haskell lib that provides basic such as `findRepo` to get the current git path https://github.com/madnight/zsh-git-prompt/commit/3f7252b0f519fb02aff5be0fea89826d1c95dd18 with is basically the same as your `find_git_root`
Today i have a look at "8. The number of stashes on the current repository."

**Comment by starcraftman** [comment]
Great thanks for looking into this. I am somewhat disappointed by python's slow startup on an older machine of mine. I did expect it to be clear. Makes the prompt between invocations slow. People will probably be happiest when we get full parity with the faster gitstatus. 

If you hadn't offered to make the patches, I probably would have rewritten `gitstatus.py` in c++ or straight shell. Both likely get us to Haskell or better performance, the latter would  be quite tedious to write the conditional logic in. I may still do one or the other, I'll think on it.

**Comment by madnight** [comment]
I think the greatest benefit regarding performance is that haskell compiles down to a small binary (direct executable assembly like c/c++) whereas python needs to invoke a full blown interpreter first and interpret the whole script on every run.

**Comment by starcraftman** [comment]
Oh yes, the majority of the time isn't even spent executing code in `gitstatus.py`.

Examples from my i7 dev box below.
```sh
$ time python -c ''
real    0m0.047s
user    0m0.030s
sys 0m0.020s

$ time git status --branch --porcelain | ./gitstatus.py 
master 0 0 0 0 0 0 0 0 origin/master 0 0
real    0m0.050s
user    0m0.035s
sys 0m0.022s

$ time git status --branch --porcelain | src/.bin/gitstatus 
master 0 0 0 0 0 0
real    0m0.009s
user    0m0.011s
sys 0m0.008s
```
